### PR TITLE
When not wrapping fallback to base64 encoding in case invalid SQS chars are being used

### PR DESCRIPTION
**Pull request by danielmarbach**
This is a follow-up on:

- https://github.com/Particular/NServiceBus.AmazonSQS/pull/2623

Which also addresses:

- https://github.com/Particular/NServiceBus.AmazonSQS/issues/2630

Originally, `DoNotWrapOutoingMessages` was designed to no longer base64 encode and wrap the message with the transport message. The idea was to purely hand over whatever we got to the transport. So in cases of raw dispatches when the message contains characters that are not compliant with the [SQS message encoding constraints](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessage.html), the transport would throw.

This has consequences for scenarios like sending metrics when the payload uses raw bytes that may be incompatible with the text restrictions of SQS. By discovering that an encoded payload has characters that are not suitable to be sent and then falling back to base64 encoding, those scenarios are handled.

The transport pump already assumes when it receives non-native messages that the body might be base4 encoded and therefore gracefully falls back to reading those encoded special cases too without any extra code required.

The consequence of the change is that when not wrapping, the payload needs to be inspected for illegal characters which has computational overhead. The current implementation uses the most straightforward approach using a source-gened regex that should be reasonably fast enough. Future optimizations could be done by checking the byte array directly for invalid byte sequences / code points.

### CODE REVIEW

File: src/NServiceBus.Transport.SQS.TransportTests.DoNotWrapOutgoingMessages/Sending_messages_with_invalid_sqs_chars.cs [hunk 1]
```diff
@@ -1,38 +1,47 @@
...
+                    return messageProcessed.SetCompleted(context);
+                },
                 (_, __) => Task.FromResult(ErrorHandleResult.Handled),
                 TransportTransactionMode.None);
 
             var headers = new Dictionary<string, string>
             {
-                { Transport.SQS.Constants.MetricsMessageMetricTypeHeaderKey, "doesn't matter" },
-                { Headers.ContentType, Transport.SQS.Constants.MetricsMessageContentTypeHeaderValue }
+                { "SomeHeader", "doesn't matter" },
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SQS.TransportTests.DoNotWrapOutgoingMessages/Sending_messages_with_invalid_sqs_chars.cs:30]: Would it make sense to add some invalid characters here as well to make this more future proof should we switch to using [message attributes](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html) in the future when the "max 10" constraint have been lifted?

File: src/NServiceBus.Transport.SQS/MessageDispatcher.cs [hunk 2]
```diff
@@ -482,6 +484,9 @@ long CalculateDelayedDeliverySeconds(UnicastTransportOperation transportOperatio
             return (body, headers);
         }
 
+        [GeneratedRegex(@"^[\u0009\u000A\u000D\u0020-\uD7FF\uE000-\uFFFD\u10000-\u10FFFF]*$", RegexOptions.Singleline)]
+        private static partial Regex ValidSqsCharacters();
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SQS/MessageDispatcher.cs:488]:
```suggestion
       static partial Regex ValidSqsCharacters();
```

File: src/NServiceBus.Transport.SQS.AcceptanceTests/Sending/When_sending_messages_with_invalid_sqs_chars.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+                return [.. byteList];
+            }
+        }
+
+        class Receiver : EndpointConfigurationBuilder
+        {
+            public Receiver() => EndpointSetup<DefaultServer>();
+
+            public class DoTheThing : Feature
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SQS.AcceptanceTests/Sending/When_sending_messages_with_invalid_sqs_chars.cs:119]:
```suggestion
            public class CatchControlMessage : Feature
```

File: src/NServiceBus.Transport.SQS/MessageDispatcher.cs [hunk 4]
```diff
@@ -17,7 +18,7 @@ namespace NServiceBus.Transport.SQS
     using Settings;
     using Transport;
 
-    class MessageDispatcher : IMessageDispatcher
+    partial class MessageDispatcher : IMessageDispatcher
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SQS/MessageDispatcher.cs:21]: Why do we need a partial here?

**Review by andreasohlund — APPROVED** [review]

File: src/NServiceBus.Transport.SQS/MessageDispatcher.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.Transport.SQS/MessageDispatcher.cs:488]: Partial methods with return types require a visibility modifier

File: src/NServiceBus.Transport.SQS/MessageDispatcher.cs (hunk 4 above)
Comment (danielmarbach) [reply src/NServiceBus.Transport.SQS/MessageDispatcher.cs:21]: The source gen for the regex requires that

**Review by lailabougria — APPROVED** [review]

**Comment by andreasohlund** [comment]
The regex adds 12ms when used so that all seems fine

image

**Pull request merged**